postgres-protocol/src/types/numeric.rs

```diff
@@ -0,0 +1,1583 @@
+/// Returns digits in most-significant-first order.
+fn build_integer_digits_from_bytes(digits: &[u8]) -> Result<Vec<i16>, ParseError> {
+    // Skip leading zeros
+    let trimmed_start = digits
+        .iter()
+        .position(|&b| b != b'0')
+        .unwrap_or(digits.len());
+
+    if trimmed_start == digits.len() {
+        return Ok(Vec::new());
+    }
+
+    let trimmed = &digits[trimmed_start..];
+
+    // Process in chunks of 4 from the right, converting directly to base-10000
+    let mut result = Vec::with_capacity(trimmed.len().div_ceil(4));
+
+    for chunk in trimmed.rchunks(4) {
+        let mut digit = 0i16;
+        for &b in chunk {
+            if !b.is_ascii_digit() {
+                return Err("invalid digit character".into());
+            }
+            digit = digit * 10 + (b - b'0') as i16;
+        }
+        result.push(digit);
+    }
+
+    result.reverse(); // Reverse to get most-significant-first order
+    Ok(result)
+}
+
+/// Builds base-10000 digits from decimal part bytes.
+fn build_decimal_digits_from_bytes(digits: &[u8]) -> Result<Vec<i16>, ParseError> {
+    if digits.is_empty() {
+        return Ok(Vec::new());
+    }
+
+    // Process in 4-digit chunks, padding the last chunk
+    let mut result = Vec::with_capacity(digits.len().div_ceil(4));
+
+    for chunk in digits.chunks(4) {
+        let mut digit = 0i16;
+        for &b in chunk {
+            if !b.is_ascii_digit() {
+                return Err("invalid digit character".into());
+            }
+            digit = digit * 10 + (b - b'0') as i16;
+        }
+
+        // Pad to 4 digits for proper base-10000 representation
+        if chunk.len() < 4 {
+            digit *= 10i16.pow((4 - chunk.len()) as u32);
+        }
+
+        result.push(digit);
+    }
+
+    Ok(result)
+}
+
+/// Normalize leading zeros
+fn normalize_leading_zeros(digits: &mut Vec<i16>, weight: &mut i16) {
+    let leading_zero_count = digits.iter().take_while(|&&d| d == 0).count();
+
+    if leading_zero_count > 0 {
+        if leading_zero_count == digits.len() {
+            // All digits are zero
+            *weight = 0;
+            digits.clear();
+        } else {
+            // Remove leading zeros
+            digits.drain(0..leading_zero_count);
+            *weight -= leading_zero_count as i16;
+        }
+    }
+}
+
+/// Remove trailing zeros
+fn normalize_trailing_zeros(digits: &mut Vec<i16>) {
+    // Find last non-zero position
+    if let Some(last_nonzero) = digits.iter().rposition(|&d| d != 0) {
+        digits.truncate(last_nonzero + 1);
+    } else {
+        digits.clear();
+    }
+}
+
+fn split_e(s: &[u8]) -> (&[u8], Option<&[u8]>) {
+    let mut s = s.splitn(2, |&b| b == b'e' || b == b'E');
+    let first = s.next().unwrap();
+    let second = s.next();
+    (first, second)
+}
+
+fn split_decimal(s: &[u8]) -> (&[u8], Option<&[u8]>) {
+    let mut s = s.splitn(2, |&b| b == b'.');
+    let first = s.next().unwrap();
```

Comment:
Quickly scrolling through the code it's non-obvious that this panic is unreachable.